table.py

```diff
@@ -377,8 +392,19 @@ def ai(cls, tables):
     @classmethod
     @ModelView.button
     def compute(cls, tables):
-        with Transaction().set_context(queue_name=QUEUE_NAME):
+        transaction = Transaction()
+        cursor = transaction.connection.cursor()
+
+        with transaction.set_context(queue_name=QUEUE_NAME):
             for table in tables:
+                message = gettext('babi.msg_table_sql_error', table=table.rec_name)
+
+                if table.type == 'query':
+                    table._drop()
+                    try:
+                        cursor.execute('CREATE VIEW "%s" AS %s' % (table.table_name, table._stripped_query))
+                    except DuplicateColumn as e:
+                        raise UserError('%s\n%s' % (message, str(e)))
```

Comment:
I think I'd leave this function as it was before this patch and occurred to me something more generic:

In the `_compute()` method, around line 461 in `except Exception as e:` we can check the status of the transaction (see https://www.psycopg.org/docs/connection.html#connection.status, https://www.psycopg.org/docs/extensions.html#transaction-status-constants). If the transaction is INERROR, then we may be able to explicitly call Transaction().connection.rollback() and then the notify() method could be called without problems.

Reply:
See changeset 73693fd6857918109b0b8375133bb7e0944a8e4a